### PR TITLE
add english document for tapable-hooks

### PR DESCRIPTION
**Pull request by stauren**
also updated server-hooks.ts files with hooks comments

### CODE REVIEW

File: docs/tapable-hooks.md [hunk 1]
```diff
@@ -5,43 +5,30 @@ permalink: /tapable-hooks.html
...
-1. 服务端缓存的使用逻辑
+It's also important that web applications may need other logics on the server during SSR such as:
+1. Fetch data from APIs used in page rendering.
+1. Caching and reusing of SSR HTML.
+1. Error handling like fail to fetch data or 404 visit.
+1. Generation of HTML fragments outside the `<App>` tag.
+1. Statistics and reporting of data.
 
-这些逻辑可以放到 Vue render bundle 中，但这会导致服务端执行代码和客户端执行代码难以分开，尤其是在他们对第三方包有不同依赖的时候，可能导致客户端加载大量不必要的服务端运行代码。另外一个选择是把业务的服务端逻辑硬编码到服务端代码中（这也是 Vise 早期的方式），但这会导致扩展性的问题。
+Some of these logics could be put in hooks of web user interface libraries such as `created`, but the downside is server logic and client logic is mixed. Sometimes they have totally different 3rd party dependencies, may cause the generated client bundle to contain unnecessary sever packages and impact user experience.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: Some of these logics could be put in hooks of web user interface libraries such as `created`, but the downside is server logic and client logic are mixed. Sometimes they have totally different 3rd party dependencies, which may cause the generated client bundle to contain unnecessary sever packages and impact user experience.

File: docs/tapable-hooks.md [hunk 2]
```diff
@@ -5,43 +5,30 @@ permalink: /tapable-hooks.html
...
-1. 通过 hooks 定义标准的、与业务无关的服务端生命周期，剥离特定业务和 server 端核心实现；
-1. 多个业务 app 可以并存于同一服务器而不冲突，各个业务 app 独立定义自身的服务端逻辑；
-1. hooks 使用灵活、可扩展，各个服务端的流程通过 hooks 标准化，服务器开发者和业务 app 开发者、插件开发者可以把各自的逻辑通过 hooks 整合到一起；
+Another way is putting those logics in HTTP servers, which is the early choice of Vise. But it's proved to cause scalability issues, since as the apps grows bigger and more apps deployed on the sever, the codes written in the HTTP servers could easily lose control.
 
-## Server Hooks 流程图
-![Vise tapable hooks 整体流程](./images/tapable-hooks.png)
+Vise defined the server life cycle, which is based on [tapable] and exposes multiple hooks from the point a server receiving an HTTP request from the user to the server sending the HTTP response of the request:
+1. Server logics of certain app belongs to the app but not the HTTP server anymore. 
+1. Server only packages do not exist in the generated client bundles.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: 这里 server-only 用连字符是不是比较好理解

File: docs/tapable-hooks.md [hunk 3]
```diff
@@ -5,43 +5,30 @@ permalink: /tapable-hooks.html
...
+1. Multiple apps could be deployed in the same server and scale up.
+1. Hooks is standard, flexible and extendable. Certain server side process could be abstracted as an Vise hooks plugin composed of several coordinated hooks and be reused easily.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: hooks are

File: docs/tapable-hooks.md [hunk 4]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-| 参数类型 | HTTPRequest |
-| 返回值类型 | Omit<RenderResult, 'type' \| 'renderBy'> \| void |
+| Name | receiveRequest |
+| Hook Type | AsyncParallelBailHook |
+| Parameters | HTTPRequest |
+| Return Value | Omit<RenderResult, 'type' \| 'renderBy'> \| void |
 
 ### hooks.requestResolved
-HTTP 请求解析完成，多个钩子函数顺序执行传递解析结果，可在此修改解析结果（注意：如果修改了 url，会导致 hydration 时候出现 mismatch：js 端看到的是修改前的 urlA，服务端看到的是修改后的 urlB，所以如果这里修改 url，需要配合前端的逻辑同步修改）
+Tapped functions receive a `ResolveRequest` and return a `ResolveRequest`. Marks could be added to context of certain requests for later process. Be careful of [hydration mismatch] if you change data in the HTTPRequest.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: ResolvedRequest

File: docs/tapable-hooks.md [hunk 5]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-| 参数类型 | CacheInfo |
-| 返回值类型 | string \| void |
+| Name | findCache |
+| Hook Type | AsyncParallelBailHook |
+| Parameters | CacheInfo |
+| Return Value | string \| void |
 
 ### hooks.hitCache
-在 HTML 缓存命中后并行执行所有钩子，然后响应 HTTP 请求，无法在此更改响应，可做统计等
+Tapped functions will be notified a successful cache hit event.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: Tapped functions will be notified with a successful cache hit event.

File: docs/tapable-hooks.md [hunk 6]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-| 参数类型 | HitCache |
-| 返回值类型 | void |
+| Name | hitCache |
+| Hook Type | AsyncParallelHook |
+| Parameters | HitCache |
+| Return Value | void |
 
 ### hooks.beforeRender
-在准备使用 Vue render bundle 等服务端渲染包生成 HTML 之前调用，可用来请求依赖数据等，多个钩子顺序执行传递请求参数
+Tapped functions will be called one by one before rendering HTML with server renderer provided by web UI libraries. Typically this could be used to fetch data for SSR. Data should be transferred in `RenderContext.extra`.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: 这里 one by one 可以改成 will be called in order before xxxxx, 感觉可读性强一点

File: docs/tapable-hooks.md [hunk 7]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-| 参数类型 | RenderContext |
-| 返回值类型 | RenderResult |
+| Name | render |
+| Hook Type | AsyncParallelBailHook |
+| Parameters | RenderContext |
+| Return Value | RenderResult |
 
 ### hooks.afterRender
-在 App 渲染完成后执行，根据渲染成功或失败，RenderResult 可能为成功或失败，如需重载渲染结果，钩子可以返回更改后的 RenderResult。渲染结果中包含 SsrBundleResult，服务端会使用 SsrBundleResult 中的值重新拼装页面模板。这里可以简单替换掉页面 template 而不引起 hydration mismatch (模板是 Vue app 以外的部分)。注意钩子瀑布流顺序执行。
+Tapped functions will be called one by one after render finishes, processing the result of a successful rendering or handle the error of a failed rendering.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: 同上

File: docs/tapable-hooks.md [hunk 8]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-| 参数类型 | RenderResult |
-| 返回值类型 | RenderResult |
+| Name | afterRender |
+| Hook Type | AsyncSeriesWaterfallHook |
+| Parameters | RenderResult |
+| Retuan Value | RenderResult |
 
 ### hooks.beforeResponse
-在所有 HTTP 响应发送前执行，任意回调函数(tapped function)返回结果即终止其它回调函数执行。优先返回 HTTPResponse 将替代原有 HTTPResponse 返回。参数 RenderResult 包含 RenderContext 中各钩子添加的 meta data 和渲染异常 Error 等信息，可通过它们构建最终响应 HTTPResponse。
+Tapped functions will be called after `afterRender`. This is the last hook before HTTP Response is sent and it's mainly used for compose the HTTP Response, which could be generated with all data in the `RenderResult` which contains the `RenderContext`.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: Tapped functions will be called after `afterRender`. This is the last hook before HTTP Response is sent and it's mainly used for composing the HTTP Response, which could be generated with all data in the `RenderResult` containing the `RenderContext`.

File: docs/tapable-hooks.md [hunk 9]
```diff
@@ -52,103 +39,119 @@ sendHTTPResponse        // 发送 HTTP 响应
...
-#### server-hooks.ts 配置文件
-一份典型的 `server-hooks.ts` 文件内容如下：
+| Name | beforeResponse |
+| Hook Type | AsyncParallelBailHook |
+| Parameters | RenderResult |
+| Return Value | HTTPResponse \| void |
+
+## Hooks Usage
+### App developer
+App developer use [Vise command line tool](./commandline-tool.html) to created scaffold of an app first, then there will be a app directory described in [App Directory Structure](./start-develop.html#app-directory-structure).
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: App developer can use [Vise command line tool](./commandline-tool.html) to create the scaffold of an app, then there will be an app directory as described in [App Directory Structure](./start-develop.html#app-directory-structure).

File: docs/tapable-hooks.md [hunk 10]
```diff
@@ -231,37 +236,38 @@ const serverHooks: ViseHooks = {
     return result;
   },
 
-  // 在 HTML 缓存命中后并行执行所有钩子，然后响应 HTTP 请求，无法在此更改响应，可做统计等
+  // Tapped functions will be notified a successful cache hit event.
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: Tapped functions will be notified with a successful cache hit event.

File: docs/tapable-hooks.md [hunk 11]
```diff
@@ -320,93 +328,94 @@ const serverHooks: ViseHooks = {
...
+- The default export value is constraint by the `ViseHooks` type, all parameter and return value types could have IDE notice.
+- Vise defined multiple data types in the lifecycle such as `HTTPRequest`, `HTTPResponse`, `RenderContext`, `ResolvedRequest`, `RenderResult` to support data transmission during multiple hooks, all these types could be imported from vise-ssr package. More detail at: [Key Data Types](./key-data-types.html)
+- A typical operation in a tapped function is to modify a small part of a complex data structure such as `RenderContext`. Vise provided `mergeConfig` util function with deep partial type support to help.
+
+### HTTP Server Developer
+#### Core Hook Classes
+Vise defined multiple hook related Classes which HTTP Server developer need to understand and use:
+- HookLifeCycle: accept `ViseHooks` config which contains all the tapped functions as parameter, defined whole SSR life cycle. The HTTP Server should create an instance with merged ViseHooks config and hand over the HTTP request to it.
+- HookManager: define core hooks data structure, hook type, tapped function I/O data type
```
Comment (yyyyccs7) [inline docs/tapable-hooks.md]: define core hook data structure, hook type, tapped function I/O data type

File: docs/tapable-hooks.md (hunk 2 above)
Comment (stauren) [reply docs/tapable-hooks.md]: make sense, ok

File: docs/tapable-hooks.md (hunk 3 above)
Comment (stauren) [reply docs/tapable-hooks.md]: OK

File: docs/tapable-hooks.md (hunk 1 above)
Comment (stauren) [reply docs/tapable-hooks.md]: OK

File: docs/tapable-hooks.md (hunk 4 above)
Comment (stauren) [reply docs/tapable-hooks.md]: All ResolveRequest are rename to ResolvedRequest now.

File: docs/tapable-hooks.md (hunk 5 above)
Comment (stauren) [reply docs/tapable-hooks.md]: Fixed.
I think we could point out the bug in an easier-to-read form, like this:

> notified a successful cache hit event 

notified **_with_** a successful cache hit event

File: docs/tapable-hooks.md (hunk 8 above)
Comment (stauren) [reply docs/tapable-hooks.md]: Fixed.
compose -> composing
which contains -> containing

File: docs/tapable-hooks.md (hunk 7 above)
Comment (stauren) [reply docs/tapable-hooks.md]: which ditto… Sorry I can't piece it together.

File: docs/tapable-hooks.md (hunk 9 above)
Comment (stauren) [reply docs/tapable-hooks.md]: Terrific, fixed.

File: docs/tapable-hooks.md (hunk 7 above)
Comment (stauren) [reply docs/tapable-hooks.md]: one by one → in order
fixed

File: docs/tapable-hooks.md (hunk 10 above)
Comment (stauren) [reply docs/tapable-hooks.md]: added `with`
fixed

File: docs/tapable-hooks.md (hunk 11 above)
Comment (stauren) [reply docs/tapable-hooks.md]: hooks → hook
fixed

File: docs/tapable-hooks.md (hunk 6 above)
Comment (stauren) [reply docs/tapable-hooks.md]: fixed

**Comment by yyyyccs7** [comment]
LGTM

**Pull request merged**